torch/distributed/_tensor/experimental/dmap.py

```diff
@@ -0,0 +1,155 @@
+            every :class:`DTensor`s to be placed on the same device mesh.
+        in_placements (Union[`PlacementType`, Tuple[`PlacementType`, ...]], optional):
+            the required placements of the input :class:`DTensor`s. If not specified,
+            this will be inferred from the input :class:`DTensor`s' device mesh.
+        resharding (bool, optional):
+            the bool value indicating whether to reshard the input :class:`DTensor`s when
+            their placements are different from the required input placements. If this value
+            is `False` and a :class:`DTensor` has a different placement, an exception will be
+            raised. Default: `False`.
+
+    Returns:
+        A `Callable` that applies `func` to each local shard of the input :class:`DTensor`
+        and returns a :class:`DTensor` constructed from the return value of `func`.
+
+    Raises:
+        AssertionError: If the input :class:`DTensor`s are not placed on the same device mesh, or
+        if they are placed on a different device mesh than the `device_mesh` argument passed in.
+
+        AssertionError: If the placements in `out_placements` for non-:class:`DTensor` return value
+        is not `None`.
+
+        ValueError: If `resharding=False` but the input :class:`DTensor` needs resharding according to
+        `in_placements`.
+
+    Example:
+        >>> # xdoctest: +SKIP("distributed")
+        >>> def mm_allreduce_forward(device_mesh, W, X):
+        >>>     partial_sum_tensor = torch.mm(W, X)
+        >>>     reduced_tensor = funcol.all_reduce(partial_sum_tensor, "sum", device_mesh)
+        >>>     return reduced_tensor
+        >>>
+        >>> W = torch.randn(12, 8, requires_grad=False)
+        >>> X = torch.randn(8, 16, requires_grad=False)
+        >>> Y = torch.mm(W, X)
+        >>> row_wise = [Shard(0)]  # row-wise sharding placements on 1-d mesh
+        >>> col_wise = [Shard(1)]  # col-wise sharding placements on 1-d mesh
+        >>>
+        >>> # d_mm_allreduce_forward is the function wrapped with DTensor/Tensor convertion
+        >>> d_mm_allreduce_forward = dmap(
+        >>>     mm_allreduce_forward,
+        >>>     out_placements=[Replicate()],
+        >>>     device_mesh=device_mesh,
+        >>>     in_placements=[col_wise, row_wise],
+        >>> )
+        >>>
+        >>> dW = distribute_tensor(W, device_mesh, col_wise)  # col-wisely sharded W tensor
+        >>> dX = distribute_tensor(X, device_mesh, row_wise)  # row-wisely sharded X tensor
+        >>> dY = d_mm_allreduce_forward(dW, dX)  # apply d_mm_allreduce_forward to DTensors
+
+    NOTE: This API is currently experimental and subject to change
+    """
+
+    def wrapped(*args, **kwargs):
+        # process input args
+        flat_args, args_spec = tree_flatten(args)
+
+        # we assume every DTensor object is placed on the same device mesh
+        flat_local_args = []
+        nonlocal device_mesh  # access var device_mesh from the outer scope
```

Comment:
what does this mean? I thought device_mesh from the dmap API should be visible here even without nonlocal?

Reply:
The reason we need `nonlocal` here is that `device_mesh` may be updated if it's passed in as `None`. Without `nonlocal`, the assignment `device_mesh = arg.device_mesh` will be treated as creating a new local variable `device_mesh` and initializing it to `arg.device_mesh`. If without any `device_mesh = ...` assignment, we would not need the `nonlocal` keyword.